### PR TITLE
Add unit tests for com.ruoyi.common.support.Convert

### PR DESCRIPTION
**Pull request by EricHetti**
I've analysed your codebase and noticed that com.ruoyi.common.support.Convert is not fully tested.
I've written some tests for the methods in this class with the help of [Diffblue Cover](https://www.diffblue.com/opensource).

Hopefully, these tests will help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other classes that you consider important.

### CODE REVIEW

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
+package com.ruoyi.common.support;
+
+import java.nio.ByteBuffer;
```
Comment (petersear) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: This import could be with the other java.* (including another for java.nio) imports.

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    public void testToNumber() {
+        assertEquals(8, toNumber(null, 8));
+        assertEquals(8, toNumber("", 8));
+        assertEquals((long) 123, toNumber("123", 8));
+        assertEquals(8, toNumber("foo", 8));
+        assertEquals(1, toNumber(1));
+    }
+
+    @Test
+    public void testToInt() {
```
Comment (petersear) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java:58]: I feel like there is an assert missing - 5 asserts for 6 returns.

**Review by petersear — CHANGES_REQUESTED** [review]

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+import static com.ruoyi.common.support.Convert.*;
+import static org.junit.Assert.*;
+
+public class ConvertTest {
+
+    @Test
+    public void testToStr() {
+        assertEquals("bar", toStr(null, "bar"));
+        assertEquals("123", toStr(123, "bar"));
+        assertEquals("foo", toStr("foo"));
```
Comment (sudo-jaa) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Out of curiosity more than anything, i'd quite like to see an Assert here for when we call the toStr method but pass in two nulls. 

The method is overloaded to revert to assigning null as the defaultValue if one isn't passed, but for redundancy it might be nice to show that null is still returned if a null value is passed for the defaultValue.

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    public void testToInt() {
+        assertEquals(8, (int) toInt(null, 8));
+        assertEquals(8, (int) toInt("", 8));
+        assertEquals(123, (int) toInt("123", 8));
+        assertEquals(8, (int) toInt("foo", 8));
+        assertEquals(1, (int) toInt(1));
+    }
+
+    @Test
+    public void testToIntArray() {
```
Comment (sudo-jaa) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java:68]: I think that instead of testing the length of the arrays crated by the to*Array methods, it would be better to to explicitly create Arrays by hand and assert equality between what you have created, and what the method has returned.

**Review by sudo-jaa — CHANGES_REQUESTED** [review]
A couple of general comments/requests

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 3 above)
Comment (petersear) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: If we call toStr(null, null) it will return null.
This is the same (in terms of testing lines) as calling toStr(null, "bar") and returning "bar" or have I missed the point?

Comment (sudo-jaa) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Indeed, the line has been covered in terms of what is actually being executed. It may be a question of testing style, but I quite like the thoroughness of reaching the null return directly, as well as indirectly with the overloaded method.

Comment (petersear) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Are you saying there is a point in returning null instead of "bar"? I could see that might be the case if passing null, as opposed to a string, would throw an exception for example.
There is no indirect returning of null in the proposed tests ie. there is no call to toStr(null).

Comment (sudo-jaa) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: I would say so, yes. We get the same return from toStr(null) as we do from toStr(null, null), but the former will get this return (or exception) by calling a separate function. 
Therefore I would probably argue that reaching this same end state by both variants of the toStr() function would result in a more robust test method.

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 2 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java:58]: Thank you. Done

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 4 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java:68]: ok thanks. Done :)

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 1 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Do you mean do this - `import java.*`
to replace -
```
import java.nio.ByteBuffer;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.charset.Charset;
import java.text.NumberFormat;
import java.text.ParseException;
import java.util.Arrays;
import java.util.Set;
import java.util.TreeSet;
```

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 3 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: So.... what am i doing for this? I got all the lines covered for the method. Do you want me to still test for the (null,null) values?

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 1 above)
Comment (petersear) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: NO! ;-)
I was just using a shorthand to suggest grouping together all imports that start with "java."
import java.nio.ByteBuffer; is out on its own. I was suggesting moving it "into the pack":

import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.ByteBuffer;  **<---**
import java.nio.charset.Charset;
import java.text.NumberFormat;
import java.text.ParseException;
import java.util.Arrays;
import java.util.Set;
import java.util.TreeSet;

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 3 above)
Comment (petersear) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: I am happy with the proposed tests which cover all the lines. I'd only be in favour of adding an extra test with a null input if it exercised the code differently.

**Review by petersear — COMMENTED** [review]
Once the imports are tidied up I think that resolves my outstanding questions.
If you could push the changes which resolve the issues marked as "done", we can check those too.

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 1 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Ok :) done

**Review by petersear — APPROVED** [review]

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    @Test
+    public void testToByte() {
+        assertEquals((byte) 8, (byte) toByte(null, (byte) 8));
+        assertEquals((byte) 1, (byte) toByte((byte) 1, (byte) 8));
+        assertEquals((byte) 8, (byte) toByte("", (byte) 8));
+        assertEquals((byte) 123, (byte) toByte("123", (byte) 8));
+        assertEquals((byte) 8, (byte) toByte("foo", (byte) 8));
+        assertEquals((byte) 1, (byte) toByte(1));
```
Comment (freddytuxworth) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: IMO instead of casting both the `expected` and the returned boxed `Byte` to `byte`, it might be better to box the `expected` like `assertEquals(new Byte(8), toByte(null, (byte) 8))`. That way the assertion verifies the exact return behaviour and also I think it's a bit clearer. Same for `toShort` and `toInt`. You might disagree though?

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        assertEquals(Arrays.asList("a", "b", "c", "d"), Arrays.asList(toStrArray("a,b,c,d")));
+    }
+
+    @Test
+    public void testToLong() {
+        assertEquals(8, (long) toLong(null, (long) 8));
+        assertEquals(1, (long) toLong((long) 1, (long) 8));
+        assertEquals(8, (long) toLong("", (long) 8));
+        assertEquals(123, (long) toLong("123", (long) 8));
+        assertEquals(8, (long) toLong("foo", (long) 8));
```
Comment (freddytuxworth) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Instead of casting to `(long)` in the second param of `toLong`, you could use a long literal (`8L`)

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        assertEquals(1, (long) toLong(1));
+    }
+
+    @Test
+    public void testToDouble() {
+        assertEquals(8, toDouble(null, (double) 8), 0);
+        assertEquals(1, toDouble((double) 1, (double) 8), 0);
+        assertEquals(8, toDouble("", (double) 8), 0);
+        assertEquals(123, toDouble("123", (double) 8), 0);
+        assertEquals(8, toDouble("foo", (double) 8), 0);
```
Comment (freddytuxworth) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Instead of casting to `(double)` in the second param of `toDouble`, you could use a double literal (`8D`)

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        assertEquals(1, toDouble(1), 0);
+    }
+
+    @Test
+    public void testToFloat() {
+        assertEquals(8, toFloat(null, (float) 8), 0);
+        assertEquals(1, toFloat((float) 1, (float) 8), 0);
+        assertEquals(8, toFloat("", (float) 8), 0);
+        assertEquals(123, toFloat("123", (float) 8), 0);
+        assertEquals(8, toFloat("foo", (float) 8), 0);
```
Comment (freddytuxworth) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Instead of casting to `(float)` in the second param of `toFloat`, you could use a double literal (`8F`)

**Review by freddytuxworth — COMMENTED** [review]
Does the title refer to a different `ByteBuffer`?

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 8 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Done

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 7 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Done

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 6 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Done

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 5 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Yes I agree. Thanks

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+    @Test
+    public void testToLongArray() {
+        assertEquals(Arrays.asList(), Arrays.asList(toLongArray(null, "")));
+        assertEquals(Arrays.asList(null, null, null, null), Arrays.asList(toLongArray("a,b,c,d")));
+    }
+
+    @Test
+    public void testToStrArray() {
+        assertEquals(Arrays.asList("a", "b", "c", "d"), Arrays.asList(toStrArray("a,b,c,d")));
```
Comment (freddytuxworth) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Instead of calling `asList` for both, change expected to `new String[] {"a", "b", "c", "d"}`

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 10]
```diff
@@ -0,0 +1,205 @@
+package com.ruoyi.common.support;
```
Comment (freddytuxworth) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java:1]: Does the PR title refer to the right `Convert` class?

**Review by freddytuxworth — CHANGES_REQUESTED** [review]

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 10 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java:1]: I missed that....  :O 
thanks

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 9 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Your right.. not sure why I did it like that

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        assertEquals(8, toNumber("", 8));
+        assertEquals(123L, toNumber("123", 8));
+        assertEquals(8, toNumber("foo", 8));
+        assertEquals(1, toNumber(1));
+    }
+
+    @Test
+    public void testToInt() throws ParseException {
+        assertEquals(new Integer("8"), toInt(null, 8));
+        assertEquals(new Integer("1"), toInt(NumberFormat.getInstance().parse("1"), 8));
```
Comment (freddytuxworth) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Replace `NumberFormat.getInstance().parse("1")` with `(Number) 1L`

**Review by freddytuxworth — CHANGES_REQUESTED** [review]

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 11 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Thanks... done :)

**Review by freddytuxworth — APPROVED** [review]

**Review by paulkennethkent — APPROVED** [review]

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    public void testToNumber() {
+        assertEquals(8, toNumber(null, 8));
+        assertEquals(8, toNumber("", 8));
+        assertEquals(123L, toNumber("123", 8));
+        assertEquals(8, toNumber("foo", 8));
+        assertEquals(1, toNumber(1));
+    }
+
+    @Test
+    public void testToInt() throws ParseException {
```
Comment (paulkennethkent) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Remove `throws ParseException`

**Review by paulkennethkent — CHANGES_REQUESTED** [review]

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.text.ParseException;
+import java.util.Set;
+import java.util.TreeSet;
+
+import static com.ruoyi.common.support.Convert.*;
+import static org.junit.Assert.*;
```
Comment (paulkennethkent) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Please don't use wildcarding for imports for `org.junit.Assert.*;`, import each one individually.

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.text.ParseException;
+import java.util.Set;
+import java.util.TreeSet;
+
+import static com.ruoyi.common.support.Convert.*;
```
Comment (paulkennethkent) [inline ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: This is almost too neat and tidy. People prefer having the method name in the test call.

**Review by paulkennethkent — CHANGES_REQUESTED** [review]

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 12 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Done

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 13 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Done

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 14 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: Done

File: ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java (hunk 3 above)
Comment (Braavos6) [reply ruoyi-common/src/test/java/com/ruoyi/common/support/ConvertTest.java]: @paulkennethkent what should I do about this?

**Review by paulkennethkent — APPROVED** [review]
